src/classes/RD2_ScheduleService.cls

```diff
@@ -207,72 +207,73 @@ public without sharing class RD2_ScheduleService {
     */
     private Date getNextDonationDateGreaterThanToday(RecurringDonationSchedule__c schedule) {
 
-        Integer frequency = Integer.valueOf(schedule.InstallmentFrequency__c);
         Date firstValid = getFirstPossibleDonationDateFromStart(schedule);
 
         if (firstValid >= currentDate) {
             return firstValid;
         }
 
-        Integer unitsBetween;
-        Integer remains;
-        Date adjusted;
+        Integer adjustedFrequency =
+            (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_WEEKLY ? 7 : 1) * Integer.valueOf(schedule.InstallmentFrequency__c);
 
-        if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY || schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_FIRST_AND_FIFTEENTH) {
+        Integer unitsBetween = getDateUnitsBetweenFirstValidAndCurrentDate(firstValid, schedule.InstallmentPeriod__c);
 
-            unitsBetween = firstValid.monthsBetween(currentDate);
-            remains = Math.mod(unitsBetween, frequency);
-            adjusted = firstValid.addMonths((unitsBetween / frequency) * frequency);
-
-            if (remains > 0) {
-                adjusted = adjusted.addMonths(frequency);
-            }
+        Date adjusted = addDateUnits(firstValid, (unitsBetween / adjustedFrequency) * adjustedFrequency, schedule.InstallmentPeriod__c);
 
-            if (adjusted < currentDate) {
-                adjusted = adjusted.addMonths(frequency);
-            }
+        if (adjusted < currentDate) {
+            adjusted = addDateUnits(adjusted, adjustedFrequency, schedule.InstallmentPeriod__c);
+        }
 
-            if (schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY) {
-                adjusted = Date.newInstance(adjusted.year(), adjusted.month(), Date.daysInMonth(adjusted.year(),adjusted.month()));
-            }
+        if (schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY && schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY) {
+            adjusted = Date.newInstance(adjusted.year(), adjusted.month(), Date.daysInMonth(adjusted.year(),adjusted.month()));
         }
 
-        else if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_YEARLY) {
+        return adjusted;
+    }
+
+    /***
+    * @description Returns the date units between an input date and current date based on installment period
+    * @param firstValid Date
+    * @param period String
+    * @return Integer
+    */
+    private Integer getDateUnitsBetweenFirstValidAndCurrentDate(Date firstValid, String period) {
+        Integer unitsBetween = 0;
 
+        if (period == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY || period == RD2_Constants.INSTALLMENT_PERIOD_FIRST_AND_FIFTEENTH) {
```

Comment:
If this method is only called from the `getNextDonationDateGreaterThanToday` method, and that method only takes an RDSchedule record, then the period will never be INSTALLMENT_PERIOD_FIRST_AND_FIFTEENTH. Does that impact the logic or the results?

Reply:
There is a call that takes a recurring donation and ends up calling `getNextDonationDateGreaterThanToday`

`    public Date getNextDonationDate(npe03__Recurring_Donation__c rd) {
        if (rd.Status__c == RD2_Constants.STATUS_CLOSED) {
            return null;
        }

        return getNextDonationDate(getNewSchedules(rd));
    }
`

Also, schedules can have 1st and 15th period.

Reply:
Ah. I see what you mean. The `RecurringDonationSchedule` record retain the period of "`st and 15th", but the DayOfMonth field on the 2 records get a 1 and 15 respectively. I though that at one point these were actually both set to Monthly. This way works fine as well though. I just mis-remembered how the records were created.